File: 270-ClosestBinarySearchTreeValue/270-ClosestBinarySearchTreeValue.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        ans = root.val
        
        def dfs(node):
            nonlocal ans
            #case when there are no nodes
            if not node:
                return
            
            if abs(node.val - target) < abs(ans - target):
                ans = node.val
            elif abs(node.val - target) == abs(ans - target): 
                ans = min(ans, node.val)

            if target < node.val:
                dfs(node.left)
            else:
                dfs(node.right)
        
        dfs(root)
        
        return ans

File: 270-ClosestBinarySearchTreeValue/270-ClosestBinarySearchTreeValue.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} target
 * @return {number}
 */
var closestValue = function(root, target) {
    ans = root.val
    const dfs = (node) => {
        if(!node) return

        if(Math.abs(node.val - target) < Math.abs(ans - target)){
            ans = node.val;
        } else if (Math.abs(node.val - target) == Math.abs(ans - target)){
            ans = Math.min(ans, node.val);
        }

        if(target < node.val){
            dfs(node.left);
        } else{
            dfs(node.right);
        }
    }
    dfs(root);

    return ans;
};